aggregator/generic_elem.go

```diff
@@ -175,16 +195,41 @@ func (e *GenericElem) ResetSetData(
 	return nil
 }
 
-// AddMetric adds a new metric value.
-func (e *GenericElem) AddMetric(timestamp time.Time, mu unaggregated.MetricUnion) error {
+// AddUnion adds a metric value union at a given timestamp.
+func (e *GenericElem) AddUnion(timestamp time.Time, mu unaggregated.MetricUnion) error {
 	alignedStart := timestamp.Truncate(e.sp.Resolution().Window).UnixNano()
-	agg, err := e.findOrCreate(alignedStart)
+	lockedAgg, err := e.findOrCreate(alignedStart)
 	if err != nil {
 		return err
 	}
-	agg.Lock()
-	agg.Add(mu)
-	agg.Unlock()
+	lockedAgg.Lock()
+	lockedAgg.aggregation.AddUnion(mu)
+	lockedAgg.Unlock()
+	return nil
+}
+
+// AddUnique adds a metric value from a given source at a given timestamp.
+// If previous values from the same source have already been added to the
+// same aggregation, the incoming value is discarded.
+func (e *GenericElem) AddUnique(timestamp time.Time, value float64, sourceID uint32) error {
+	alignedStart := timestamp.Truncate(e.sp.Resolution().Window).UnixNano()
+	lockedAgg, err := e.findOrCreate(alignedStart)
+	if err != nil {
+		return err
+	}
+	lockedAgg.Lock()
```

Comment:
See above for the rationale.

I could also rename it to just `agg` but it'll be `agg.aggregation.blah` instead of `lockedAgg.aggregation.blah` when referencing some of the nested data fields. Plus `lockedAgg` makes it more explicit that this is protected by a lock. Wdyt?

Reply:
Sure, I'm fine either way